parity-clib/src/java.rs

```diff
@@ -15,34 +15,31 @@
 // along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.
 
 use std::{mem, ptr};
+use std::ffi::c_void;
 use std::sync::Arc;
-use std::time::Duration;
-use std::thread;
-use std::os::raw::c_void;
 
-use {parity_config_from_cli, parity_destroy, parity_start, parity_unsubscribe_ws, ParityParams, error};
+use {Callback, parity_config_from_cli, parity_destroy, parity_start, parity_rpc_worker, parity_unsubscribe_ws,
+	parity_ws_worker, ParityParams};
 
-use futures::{Future, Stream};
-use futures::sync::mpsc;
 use jni::{JavaVM, JNIEnv};
 use jni::objects::{JClass, JString, JObject, JValue, GlobalRef};
 use jni::sys::{jlong, jobjectArray, va_list};
-use tokio_current_thread::CurrentThread;
-use parity_ethereum::{RunningClient, PubSubSession};
+use parity_ethereum::RunningClient;
 
 type CheckedQuery<'a> = (&'a RunningClient, String, JavaVM, GlobalRef);
 
 // Creates a Java callback to a static method named `void callback(Object)`
-struct Callback<'a> {
+struct JavaCallback<'a> {
 	jvm: JavaVM,
 	callback: GlobalRef,
 	method_name: &'a str,
 	method_descriptor: &'a str,
 }
 
-unsafe impl<'a> Send for Callback<'a> {}
-unsafe impl<'a> Sync for Callback<'a> {}
-impl<'a> Callback<'a> {
+unsafe impl<'a> Send for JavaCallback<'a> {}
+unsafe impl<'a> Sync for JavaCallback<'a> {}
```

Comment:
It wasn't introduced by this PR, but could you explain why it's safe for [`GlobalRef`](https://docs.rs/jni/0.10.2/jni/objects/struct.GlobalRef.html)?

Reply:
I don't understand what you mean. Can you elaborate a bit?

Reply:
Sorry, I meant why is it safe to impl Sync and Send for JavaCallback?

Reply:
That is a good question I have to admit I'm not completely sure (I more or less copy-pasted it from the `lib.rs`) but I argue that `JavaVM` and `GlobalRef` is never mutated and the `Java code` is thread-safe!

Also both [JavaVM](https://github.com/jni-rs/jni-rs/blob/master/src/wrapper/java_vm/vm.rs#L13-#L18) [GlobalRef](https://github.com/jni-rs/jni-rs/blob/master/src/wrapper/objects/global_ref.rs#L28-#L66) have unsafe `Send + Sync` impls but an inner-type `GlobalRefGuard` doesn't with the unsafe impls I get the following error:

```bash
53 | impl<'a> Callback for JavaCallback<'a> {
   |          ^^^^^^^^ `*mut jni::sys::_jobject` cannot be shared between threads safely
   |
   = help: within `jni::objects::global_ref::GlobalRefGuard`, the trait `std::marker::Sync` is not implemented for `*mut jni::sys::_jobject`
   = note: required because it appears within the type `jni::objects::JObject<'static>`
   = note: required because it appears within the type `jni::objects::global_ref::GlobalRefGuard`
   = note: required because of the requirements on the impl of `std::marker::Sync` for `std::sync::Arc<jni::objects::global_ref::GlobalRefGuard>`
   = note: required because it appears within the type `jni::objects::GlobalRef`
   = note: required because it appears within the type `java::JavaCallback<'a>`
```
Thoughts?

Reply:
The reason I asked, is because the [docs for GlobalRef 0.10.2](https://docs.rs/jni/0.10.2/jni/objects/struct.GlobalRef.html) showed 
``` rust
impl !Sync for GlobalRef 
```
It seems like `Sync` was added later (https://github.com/jni-rs/jni-rs/pull/145), and the implementation hasn't changed for `GlobalRef`.

Reply:
Thanks good point let’s ask them to bump the version so we can get rid of the unsafe part then :)

Reply:
cc @DarkEld3r